src/D2L.Security.OAuth2/Keys/KeyManagementService.cs

```diff
@@ -0,0 +1,207 @@
+	internal sealed class KeyManagementService : IKeyManagementService, IPrivateKeyProvider, IDisposable {
+		private readonly IPublicKeyDataProvider m_publicKeys;
+		private readonly IPrivateKeyDataProvider m_privateKeys;
+		private readonly IDateTimeProvider m_clock;
+		private readonly OAuth2Configuration m_config;
+
+		private D2LSecurityToken m_current = null;
+
+		public KeyManagementService(
+			IPublicKeyDataProvider publicKeys,
+			IPrivateKeyDataProvider privateKeys,
+			IDateTimeProvider clock,
+			OAuth2Configuration config
+		) {
+			m_publicKeys = publicKeys;
+			m_privateKeys = privateKeys;
+			m_clock = clock;
+			m_config = config;
+		}
+
+		async Task<D2LSecurityToken> IPrivateKeyProvider.GetSigningCredentialsAsync() {
+			var current = Volatile.Read( ref m_current );
+
+			var now = m_clock.UtcNow;
+
+			if ( current == null || current.ValidTo <= now ) {
+				// Slow path: RefreshKeyAsync() wasn't called on boot and/or it
+				// isn't being called in a background job.
+				await RefreshKeyAsync( current, now )
+					.ConfigureAwait( false );
+
+				current = Volatile.Read( ref m_current );
+			}
+
+			return current.Ref();
+		}
+
+		async Task<TimeSpan> IKeyManagementService.RefreshKeyAsync() {
+			var current = Volatile.Read( ref m_current );
+
+			var now = m_clock.UtcNow;
+
+			await RefreshKeyAsync(
+				current,
+				now
+			).ConfigureAwait( false );
+
+			current = Volatile.Read( ref m_current );
+
+			var expectedNextRotation = current.ValidTo - m_config.KeyRotationBuffer;
+
+			if( now > expectedNextRotation ) {
+				// If that's in the past, use a short retry window.
+				return TimeSpan.FromSeconds( 10 );
+			} else {
+				// Otherwise use that but with a little buffer for key
+				// generation time/imprecisely scheduled cron jobs.
+				return expectedNextRotation.AddMinutes( 1 ) - now;
+			}
+		}
+
+		async Task IKeyManagementService.GenerateNewKeyIfNeededAsync() {
+			var now = m_clock.UtcNow;
+
+			var keys = await m_privateKeys.GetAllAsync(
+				validUntilAtLeast: now
+			).ConfigureAwait( false );
+
+			foreach( var key in keys ) {
+				if( now < key.ExpiresAt - m_config.KeyRotationBuffer ) {
+					// This key is fine, do nothing.
+					return;
+				}
+			}
+
+			var keyId = Guid.NewGuid();
+			var keyIdStr = keyId.ToString();
+
+			using var csp = new RSACryptoServiceProvider {
+				PersistKeyInCsp = false
+			};
+
+			var pub = csp.ExportParameters(
+				includePrivateParameters: false
+			);
+
+			var priv = csp.ExportCspBlob(
+				includePrivateParameters: true
+			);
+
+			var expiresAt = now + m_config.KeyLifetime;
+
+			await Task.WhenAll(
+				m_publicKeys.SaveAsync(
+					keyId,
+					new RsaJsonWebKey(
+						keyIdStr,
+						expiresAt,
+						pub
+					)
+				),
+				m_privateKeys.SaveAsync(
+					new PrivateKeyData(
+						id: keyIdStr,
+						kind: PrivateKeyData.KeyKinds.Rsa,
+						data: priv,
+						createdAt: now,
+						notBefore: now + m_config.KeyTimeUntilUse,

```

Comment:
If we didn't have a key / are recovering, I believe we won't want the `+ KeyTimeUntilUse`? (i.e. we want to start using it right away)

Reply:
That's handled more generally by `ChooseKey`, rather than at generation time.